In [65]:
#importing Libraries
import pandas as pd
import numpy as num
import os

In [66]:
path = r'C:\Users\ACER NITRO\Documents\Achievement 4'

In [67]:
df_ords_prods_merged= pd.read_pickle(os.path.join(path,'Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))

In [68]:
df_ords_prods_merged=df_ords_prods_merged.drop(columns=['busiest_day'])

In [69]:
df_ords_prods_merged.shape

(32404859, 18)

In [70]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      Busiest_days     Busiest_day busiest_period_of_day  
0   Regularly busy  Regularly busy        Average orders  
1  Least busy days  Regularly busy        Average orders  
2  Least busy days  Regularly busy           Most orders  
3  Least busy days      Least busy        Average orders  
4  Least busy days      Least busy           Most orders

## Step 2: Find the aggregated mean

In [71]:
df_ords_prods_merged.groupby('department_id').agg({'order_number':['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

## Step 3: Analyze the result. How do the results for the entire dataframe differ from those of the subset

###  The minimum mean order is 11.3 in the subset while it is 15.2 in the dataframe, and the maximum mean is 19.5 in the subset, 20.1 in the dataframe (not including department 21, the "missing" category).The department ID with the lhighest number of average order numbers is missing in subset, indicating a large number of unspecified values for the column "department_id." There is less variability in the mean orders when calculated using the entire dataframe as comparat to the smaller subset:

## Step4: create a loyalty flag for existing customers using the transform() and loc() functions.

In [72]:
#aggregating data with transform
df_ords_prods_merged['max_order'] = df_ords_prods_merged.groupby('user_id')['order_number'].transform(num.max)

In [73]:
df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   

  product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0         Soda        77              7     9.0   both  Mid-range product   
1         Soda        77              7     9.0   both  Mid-range product   
2         Soda        77              7     9.0   both  Mid-range product   
3         Soda        77              7     9.0   both  Mid-range product   
4         Soda        77              7     9.0   both  Mid-range product   

      Busiest_days     Busiest_day busiest_period_of_day  max_order  
0   Regularly busy  Regularly busy        Average orders         10  
1  Least busy days  Regularly busy        Average orders         10  
2  Least busy days  Regularly busy           Most orders         10  
3  Least busy days      Least busy        Average orders         10  
4  Least busy days      Least busy           Most orders         10

In [74]:
#Deriving columns with loc
df_ords_prods_merged.loc[df_ords_prods_merged['max_order']>40, 'loyalty_flag'] = 'loyal_customer'

In [75]:
df_ords_prods_merged.loc[(df_ords_prods_merged['max_order'] <= 40) & (df_ords_prods_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [76]:
df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [77]:
df_ords_prods_merged['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
loyal_customer      10284093
New customer         6243990
Name: count, dtype: int64

## Step5: check the basic statistics of the product prices for each loyalty category

In [78]:
df_ords_prods_merged.groupby('loyalty_flag').agg({'prices': ['mean']})

prices
                       mean
loyalty_flag               
New customer      13.294670
Regular customer  12.495717
loyal_customer    10.386336

## Step 6:  Create a spending flag for each user based on the average price across all their orders

In [79]:
#aggregating data with transform
df_ords_prods_merged['mean_price'] = df_ords_prods_merged.groupby('user_id')['prices'].transform(num.mean)

In [80]:
df_ords_prods_merged.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

In [81]:
#Deriving columns with loc
df_ords_prods_merged.loc[df_ords_prods_merged['mean_price']<10, 'spending_flag'] = 'Low_spender'


In [82]:
df_ords_prods_merged.loc[df_ords_prods_merged['mean_price']>10, 'spending_flag'] = 'High_spender'

In [83]:
df_ords_prods_merged['spending_flag'].value_counts(dropna = False)

spending_flag
Low_spender     31770614
High_spender      633864
NaN                  381
Name: count, dtype: int64

## Step7: Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column

In [84]:
# create "median_days" for aggregation results
df_ords_prods_merged['median_days'] = df_ords_prods_merged.groupby(['user_id'])['days_since_prior_order'].transform(num.median)

In [85]:
df_ords_prods_merged.head(100)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   
15   2293453       19             2                   5                 14   
16   1973799       19             5                   6                 12   
17    532817       19             7                   4                 17   
18   1573906       21            10                   3                 10   
19   1593000       31            10                   3                  8   
20   2231262       31            17                   3                 11   
21   2580647       43             6                   4                 16   
22   2187180       43             9                   4                 12   
23   2497897       52             1                   1                  9   
24   1318871       52             2                   1                 10   
25   1261384       52             3                   1                 10   
26    580568       52             5                   2                 10   
27   2428073       52             7                   2                  7   
28    342306       52             9                   1                 11   
29   1520353       52            10                   2                 10   
30   2542086       52            11                   2                  9   
31    944694       52            12                   1                 16   
32   1268191       52            15                   2                 11   
33   1122089       52            17                   2                 10   
34   1498922       52            18                   2                 10   
35    180919       52            19                   3                  8   
36   2695875       52            21                   5                 10   
37   3205960       67             1                   3                 11   
38    970922       67             2                   1                 15   
39    596585       67             3                   1                 13   
40   2037087       67             4                   2                  9   
41   2474111       67             5                   3                 14   
42   1568481       67             6                   3                 12   
43   2048527       67             8                   2                 11   
44   2389031       67             9                   2                 14   
45   2378562       67            10                   1                 10   
46   1515698       67            11                   1                 11   
47   1928050       67            12                   2                 12   
48   1949274       67            13                   1                 10   
49    551869       67            14                   1                  9   
50    768572       67     

In [86]:
# Create flag using loc
df_ords_prods_merged.loc[df_ords_prods_merged['median_days'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [87]:
df_ords_prods_merged.loc[(df_ords_prods_merged['median_days'] > 10) & (df_ords_prods_merged['median_days'] <= 20), 'frequency_flag'] = 'Regular customer'

In [88]:
df_ords_prods_merged.loc[df_ords_prods_merged['median_days'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [89]:
# Check frequencies of spending flags for accuracy
df_ords_prods_merged['frequency_flag'].value_counts(dropna=False)

frequency_flag
Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: count, dtype: int64

In [90]:

pd.options.display.max_rows = None

In [93]:
df_ords_prods_merged.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   
5   3367565        1             6                   2                  7   
6    550135        1             7                   1                  9   
7   3108588        1             8                   1                 14   
8   2295261        1             9                   1                 16   
9   2550362        1            10                   4                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN       196.0                1.0        0.0   
1                    15.0       196.0                1.0        1.0   
2                    21.0       196.0                1.0        1.0   
3                    29.0       196.0                1.0        1.0   
4                    28.0       196.0                1.0        1.0   
5                    19.0       196.0                1.0        1.0   
6                    20.0       196.0                1.0        1.0   
7                    14.0       196.0                2.0        1.0   
8                     0.0       196.0                4.0        1.0   
9                    30.0       196.0                1.0        1.0   

  product_name  ...    price_range_loc     Busiest_days     Busiest_day  \
0         Soda  ...  Mid-range product   Regularly busy  Regularly busy   
1         Soda  ...  Mid-range product  Least busy days  Regularly busy   
2         Soda  ...  Mid-range product  Least busy days  Regularly busy   
3         Soda  ...  Mid-range product  Least busy days      Least busy   
4         Soda  ...  Mid-range product  Least busy days      Least busy   
5         Soda  ...  Mid-range product   Regularly busy  Regularly busy   
6         Soda  ...  Mid-range product     Busiest days  Regularly busy   
7         Soda  ...  Mid-range product     Busiest days  Regularly busy   
8         Soda  ...  Mid-range product     Busiest days  Regularly busy   
9         Soda  ...  Mid-range product  Least busy days      Least busy   

  busiest_period_of_day max_order  loyalty_flag mean_price spending_flag  \
0        Average orders        10  New customer   6.367797   Low_spender   
1        Average orders        10  New customer   6.367797   Low_spender   
2           Most orders        10  New customer   6.367797   Low_spender   
3        Average orders        10  New customer   6.367797   Low_spender   
4           Most orders        10  New customer   6.367797   Low_spender   
5        Average orders        10  New customer   6.367797   Low_spender   
6           Most orders        10  New customer   6.367797   Low_spender   
7           Most orders        10  New customer   6.367797   Low_spender   
8           Most orders        10  New customer   6.367797   Low_spender   
9        Average orders        10  New customer   6.367797   Low_spender   

   median_days         frequency_flag  
0         20.5  Non-frequent customer  
1         20.5  Non-frequent customer  
2         20.5  Non-frequent customer  
3         20.5  Non-frequent customer  
4         20.5  Non-frequent customer  
5         20.5  Non-frequent customer  
6         20.5  Non-frequent customer  
7         20.5  Non-frequent customer  
8         20.5  Non-frequent customer  
9         20.5  Non-frequent customer  

[10 rows x 24 columns]

In [92]:
#checking the null values
df_ords_prods_merged[df_ords_prods_merged['frequency_flag'].isnull()]


order_id  user_id  order_number  orders_day_of_week  \
13645692    895835   159838             1                   0   
17251990    895835   159838             1                   0   
17622767    895835   159838             1                   0   
24138593    895835   159838             1                   0   
25880002    895835   159838             1                   0   

          order_hour_of_day  days_since_prior_order  product_id  \
13645692                 17                     NaN     10749.0   
17251990                 17                     NaN     33401.0   
17622767                 17                     NaN     23695.0   
24138593                 17                     NaN     21334.0   
25880002                 17                     NaN     22198.0   

          add_to_cart_order  reordered  \
13645692                3.0        0.0   
17251990                6.0        0.0   
17622767                2.0        0.0   
24138593                5.0        0.0   
25880002                1.0        0.0   

                                               product_name  ...  \
13645692                            Organic Red Bell Pepper  ...   
17251990                               Goat Cheese Crumbles  ...   
17622767                           California Veggie Burger  ...   
24138593                              Organic Peeled Garlic  ...   
25880002  4X Ultra Concentrated Natural Laundry Detergen...  ...   

            price_range_loc  Busiest_days  Busiest_day busiest_period_of_day  \
13645692  Mid-range product  Busiest days  Busiest day        Average orders   
17251990  Mid-range product  Busiest days  Busiest day        Average orders   
17622767  Low-range product  Busiest days  Busiest day        Average orders   
24138593  Mid-range product  Busiest days  Busiest day        Average orders   
25880002  Low-range product  Busiest days  Busiest day        Average orders   

         max_order  loyalty_flag mean_price spending_flag  median_days  \
13645692         1  New customer       7.42   Low_spender          NaN   
17251990         1  New customer       7.42   Low_spender          NaN   
17622767         1  New customer       7.42   Low_spender          NaN   
24138593         1  New customer       7.42   Low_spender          NaN   
25880002         1  New customer       7.42   Low_spender          NaN   

         frequency_flag  
13645692            NaN  
17251990            NaN  
17622767            NaN  
24138593            NaN  
25880002            NaN  

[5 rows x 24 columns]

## Step9: Export your dataframe as a pickle file

In [94]:
df_ords_prods_merged.to_pickle(os.path.join(path, 'Data','Prepared Data', 'orders_products_aggregated.pkl'))